src/main/java/rx/observers/TestSubscriber.java

```diff
@@ -271,17 +318,74 @@ public void requestMore(long n) {
      *          if the sequence of items observed does not exactly match {@code items}
      */
     public void assertReceivedOnNext(List<T> items) {
-        testObserver.assertReceivedOnNext(items);
+        if (values.size() != items.size()) {
+            assertionError("Number of items does not match. Provided: " + items.size() + "  Actual: " + values.size()
+            + ".\n"
+            + "Provided values: " + items
+            + "\n"
+            + "Actual values: " + values
+            + "\n");
+        }
+
+        for (int i = 0; i < items.size(); i++) {
+            T expected = items.get(i);
+            T actual = values.get(i);
+            if (expected == null) {
+                // check for null equality
+                if (actual != null) {
+                    assertionError("Value at index: " + i + " expected to be [null] but was: [" + actual + "]\n");
+                }
+            } else if (!expected.equals(actual)) {
+                assertionError("Value at index: " + i 
+                        + " expected to be [" + expected + "] (" + expected.getClass().getSimpleName() 
+                        + ") but was: [" + actual + "] (" + (actual != null ? actual.getClass().getSimpleName() : "null") + ")\n");
+
+            }
+        }
     }
 
     /**
+     * Wait until the current committed value count is less than the expected amount
+     * by sleeping 1 unit at most timeout times and return true if at least
+     * the required amount of onNext values have been received.
+     * @param expected the expected number of onNext events
+     * @param timeout the time to wait for the events
+     * @param unit the time unit of waiting
+     * @return true if the expected number of onNext events happened
+     * @throws InterruptedException if the sleep is interrupted
+     * @since (if this graduates from Experimental/Beta to supported, replace this parenthetical with the release number)
+     */
+    @Experimental
+    public final boolean awaitValueCount(int expected, long timeout, TimeUnit unit) throws InterruptedException {
+        while (timeout != 0 && valueCount < expected) {
+            unit.sleep(1);
```

Comment:
Instead of sleeping for 1 unit, why not using an exponential backoff capped at 1 unit.
This will ensure, fast await when the value is quickly ready, without wasting too much CPU resources.
i.e. something like that:
```
long start = System.nanoTime();
long timeoutNano = unit.convert(timeout, TimeUnit.NANOSECONDS);
long sleepPeriod = 1L;
while (valueCount < expected && System.nanoTime() - start < timeoutNano) {
    TimeUnit.NANOSECONDS.sleep(sleepPeriod);
    sleepPeriod = Math.min(sleepPeriod * 2, TimeUnit.NANOSECONDS.convert(1L, unit));
}
```